File: PyQTDB/widgets/kpi_bar.py
"""Top KPI strip: connection status, counts, closest/highest aircraft."""
from __future__ import annotations

from typing import Any, Optional

from PyQt6.QtCore import Qt, QTimer, QRectF
from PyQt6.QtGui import QColor, QFont, QPainter
from PyQt6.QtWidgets import QFrame, QHBoxLayout, QLabel, QSizePolicy, QVBoxLayout, QWidget


class _StatusLED(QWidget):
    def __init__(self, parent: Optional[QWidget] = None) -> None:
        super().__init__(parent)
        self.setFixedSize(14, 14)
        self._state = "unknown"    # "ok" | "warn" | "err" | "unknown"
        self._pulse = 0.0
        self._timer = QTimer(self)
        self._timer.setInterval(60)
        self._timer.timeout.connect(self._tick)
        self._timer.start()

    def set_state(self, state: str) -> None:
        self._state = state
        self._pulse = 1.0

    def _tick(self) -> None:
        if self._pulse > 0:
            self._pulse = max(0.0, self._pulse - 0.05)
            self.update()

    def paintEvent(self, _e) -> None:
        p = QPainter(self)
        p.setRenderHint(QPainter.RenderHint.Antialiasing, True)
        color = {
            "ok":   QColor("#5cd865"),
            "warn": QColor("#f5b642"),
            "err":  QColor("#e57373"),
            "unknown": QColor("#7c8792"),
        }[self._state]
        # halo
        if self._pulse > 0:
            halo = QColor(color)
            halo.setAlphaF(0.35 * self._pulse)
            p.setBrush(halo)
            p.setPen(Qt.PenStyle.NoPen)
            p.drawEllipse(self.rect().center(), 7 + int(4 * self._pulse),
                          7 + int(4 * self._pulse))
        p.setBrush(color)
        p.setPen(Qt.PenStyle.NoPen)
        p.drawEllipse(self.rect().center(), 5, 5)


class KpiTile(QFrame):
    def __init__(self, title: str, parent: Optional[QWidget] = None):
        super().__init__(parent)
        self.setFrameShape(QFrame.Shape.StyledPanel)
        self.setSizePolicy(QSizePolicy.Policy.Expanding, QSizePolicy.Policy.Fixed)
        self.setStyleSheet(
            "QFrame { background:#152030; border:1px solid #23324a; border-radius:6px; }"
        )
        title_lbl = QLabel(title.upper(), self)
        title_lbl.setStyleSheet("color:#7c8ea3; font-size:9pt; letter-spacing:1px;")
        self._value_lbl = QLabel("—", self)
        f = QFont()
        f.setPointSize(16)
        f.setBold(True)
        self._value_lbl.setFont(f)
        self._value_lbl.setStyleSheet("color:#e8f0f8;")
        self._sub_lbl = QLabel("", self)
        self._sub_lbl.setStyleSheet("color:#9fb0c2; font-size:8pt;")

        lay = QVBoxLayout(self)
        lay.setContentsMargins(10, 6, 10, 8)
        lay.setSpacing(0)
        lay.addWidget(title_lbl)
        lay.addWidget(self._value_lbl)
        lay.addWidget(self._sub_lbl)

    def set_value(self, value: str, sub: str = "") -> None:
        self._value_lbl.setText(value)
        self._sub_lbl.setText(sub)


class KpiBar(QWidget):
    def __init__(self, parent: Optional[QWidget] = None) -> None:
        super().__init__(parent)
        self._led = _StatusLED(self)
        self._status_text = QLabel("waiting for backend…", self)
        self._status_text.setStyleSheet("color:#c7d2df;")

        self._tile_count   = KpiTile("Aircraft in range")
        self._tile_closest = KpiTile("Closest")
        self._tile_highest = KpiTile("Highest")
        self._tile_latency = KpiTile("Backend latency")

        top = QHBoxLayout()
        top.setContentsMargins(0, 0, 0, 0)
        top.setSpacing(6)
        top.addWidget(self._led)
        top.addWidget(self._status_text, 1)

        tiles = QHBoxLayout()
        tiles.setContentsMargins(0, 0, 0, 0)
        tiles.setSpacing(8)
        for tile in (self._tile_count, self._tile_closest,
                     self._tile_highest, self._tile_latency):
            tiles.addWidget(tile)

        lay = QVBoxLayout(self)
        lay.setContentsMargins(8, 8, 8, 4)
        lay.setSpacing(6)
        lay.addLayout(top)
        lay.addLayout(tiles)

    # ------------------------------------------------------------------ API
    def set_ok(self, elapsed_ms: float, messages: list[str]) -> None:
        if messages:
            self._led.set_state("warn")
            self._status_text.setText("connected — " + "; ".join(messages))
        else:
            self._led.set_state("ok")
            self._status_text.setText(f"connected — updated in {elapsed_ms:.0f} ms")
        self._tile_latency.set_value(f"{elapsed_ms:.0f} ms", "round-trip to backend")

    def set_error(self, msg: str) -> None:
        self._led.set_state("err")
        self._status_text.setText(f"backend error — {msg}")

    def set_flights(self, flights: list[dict[str, Any]]) -> None:
        self._tile_count.set_value(str(len(flights)), "current poll")
        if not flights:
            self._tile_closest.set_value("—")
            self._tile_highest.set_value("—")
            return
        closest = min(flights, key=lambda f: _num(f.get("distance_from_obs"), 1e9))
        highest = max(flights, key=lambda f: _num(f.get("altitude"), -1))
        self._tile_closest.set_value(
            f"{_num(closest.get('distance_from_obs'), 0):.1f} nm",
            f"{closest.get('callsign', '?')}"
            f" · el {_num(closest.get('elevation'), 0):.0f}°"
        )
        self._tile_highest.set_value(
            f"{int(_num(highest.get('altitude'), 0)):,} ft",
            f"{highest.get('callsign', '?')}"
            f" · {int(_num(highest.get('speed'), 0)):,} kt"
        )


def _num(x: Any, default: float) -> float:
    try:
        return float(x)
    except (TypeError, ValueError):
        return default

File: VCQtdb/src/KpiBar.cpp
#include "KpiBar.h"

#include <QFont>
#include <QHBoxLayout>
#include <QPainter>
#include <QSizePolicy>
#include <QVBoxLayout>
#include <algorithm>

// --- StatusLED ------------------------------------------------------------
StatusLED::StatusLED(QWidget *parent) : QWidget(parent) {
    setFixedSize(14, 14);
    m_timer.setInterval(60);
    connect(&m_timer, &QTimer::timeout, this, [this]{
        if (m_pulse > 0) { m_pulse = std::max(0.0, m_pulse - 0.05); update(); }
    });
    m_timer.start();
}
void StatusLED::setState(State s) { m_state = s; m_pulse = 1.0; update(); }

void StatusLED::paintEvent(QPaintEvent *) {
    QPainter p(this);
    p.setRenderHint(QPainter::Antialiasing, true);
    QColor c;
    switch (m_state) {
        case Ok:    c = QColor("#5cd865"); break;
        case Warn:  c = QColor("#f5b642"); break;
        case Err:   c = QColor("#e57373"); break;
        default:    c = QColor("#7c8792"); break;
    }
    if (m_pulse > 0) {
        QColor h = c; h.setAlphaF(0.35 * m_pulse);
        p.setBrush(h);
        p.setPen(Qt::NoPen);
        const int r = 7 + int(4 * m_pulse);
        p.drawEllipse(rect().center(), r, r);
    }
    p.setBrush(c);
    p.setPen(Qt::NoPen);
    p.drawEllipse(rect().center(), 5, 5);
}

// --- KpiTile --------------------------------------------------------------
KpiTile::KpiTile(const QString &title, QWidget *parent) : QFrame(parent) {
    setFrameShape(QFrame::StyledPanel);
    setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);
    setStyleSheet("QFrame { background:#152030; border:1px solid #23324a; border-radius:6px; }");

    auto *t = new QLabel(title.toUpper(), this);
    t->setStyleSheet("color:#7c8ea3; font-size:9pt; letter-spacing:1px;");

    m_value = new QLabel("—", this);
    QFont f; f.setPointSize(16); f.setBold(true);
    m_value->setFont(f);
    m_value->setStyleSheet("color:#e8f0f8;");

    m_sub = new QLabel("", this);
    m_sub->setStyleSheet("color:#9fb0c2; font-size:8pt;");

    auto *lay = new QVBoxLayout(this);
    lay->setContentsMargins(10, 6, 10, 8);
    lay->setSpacing(0);
    lay->addWidget(t);
    lay->addWidget(m_value);
    lay->addWidget(m_sub);
}
void KpiTile::setValue(const QString &v, const QString &s) {
    m_value->setText(v);
    m_sub->setText(s);
}

// --- KpiBar ---------------------------------------------------------------
KpiBar::KpiBar(QWidget *parent) : QWidget(parent) {
    m_led = new StatusLED(this);
    m_statusText = new QLabel("waiting for backend…", this);
    m_statusText->setStyleSheet("color:#c7d2df;");

    m_count   = new KpiTile("Aircraft in range");
    m_closest = new KpiTile("Closest");
    m_highest = new KpiTile("Highest");
    m_latency = new KpiTile("Backend latency");

    auto *top = new QHBoxLayout();
    top->setContentsMargins(0,0,0,0);
    top->setSpacing(6);
    top->addWidget(m_led);
    top->addWidget(m_statusText, 1);

    auto *tiles = new QHBoxLayout();
    tiles->setContentsMargins(0,0,0,0);
    tiles->setSpacing(8);
    tiles->addWidget(m_count);
    tiles->addWidget(m_closest);
    tiles->addWidget(m_highest);
    tiles->addWidget(m_latency);

    auto *lay = new QVBoxLayout(this);
    lay->setContentsMargins(8,8,8,4);
    lay->setSpacing(6);
    lay->addLayout(top);
    lay->addLayout(tiles);
}

void KpiBar::setOk(double elapsedMs, const QStringList &messages) {
    if (!messages.isEmpty()) {
        m_led->setState(StatusLED::Warn);
        m_statusText->setText("connected — " + messages.join("; "));
    } else {
        m_led->setState(StatusLED::Ok);
        m_statusText->setText(QString("connected — updated in %1 ms")
                              .arg(int(elapsedMs)));
    }
    m_latency->setValue(QString("%1 ms").arg(int(elapsedMs)), "round-trip to backend");
}

void KpiBar::setError(const QString &msg) {
    m_led->setState(StatusLED::Err);
    m_statusText->setText("backend error — " + msg);
}

void KpiBar::setFlights(const QList<Flight> &flights) {
    m_count->setValue(QString::number(flights.size()), "current poll");
    if (flights.isEmpty()) {
        m_closest->setValue("—");
        m_highest->setValue("—");
        return;
    }
    const Flight *closest = &flights.first();
    const Flight *highest = &flights.first();
    for (const Flight &f : flights) {
        if (f.distanceNm < closest->distanceNm) closest = &f;
        if (f.altitudeFt > highest->altitudeFt) highest = &f;
    }
    m_closest->setValue(
        QString::number(closest->distanceNm, 'f', 1) + " nm",
        QString("%1 · el %2°")
            .arg(closest->callsign.isEmpty() ? QStringLiteral("?") : closest->callsign)
            .arg(int(closest->elevationDeg))
    );
    m_highest->setValue(
        QString::number(qint64(highest->altitudeFt)) + " ft",
        QString("%1 · %2 kt")
            .arg(highest->callsign.isEmpty() ? QStringLiteral("?") : highest->callsign)
            .arg(int(highest->speedKt))
    );
}
